pkg/planner/core/collect_column_stats_usage.go

```diff
@@ -159,6 +167,10 @@ func (c *columnStatsUsageCollector) collectPredicateColumnsForUnionAll(p *Logica
 }
 
 func (c *columnStatsUsageCollector) addHistNeededColumns(ds *DataSource) {
+	c.visitedPhysTblIDs.Insert(int(ds.physicalTableID))
```

Comment:
If we insert the id into the map after the line178 and check `if len(columns) == 0 then add it to the map`. We don't need the struct `physTblIDsWithNeededCols` in the `plan_stats.go`'s `optimize()`?

Reply:
Yes.
But personally, I tend to output information for possibly more general usage and keep the "determinate mode" specific logic out of this collector.